### PR TITLE
[DRAFT] Inroduce one-cmds python interface

### PR DESCRIPTION
**Pull request by mhs4670go**
This draft is to introduce python interface to one-cmds.

Related test will be added soon.

Related : #4278 
ONE-DCO-1.0-Signed-off-by: seongwoo <mhs4670go@naver.com>

### CODE REVIEW

**Comment by seanshpark** [comment]
Although all commands are python scripts, I've removed `.py` and wanted to act as a program without extension.
Is there any particular reason to add `.py` extension?

File: compiler/one-cmds/one_build.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        action='store_true',
+        help='show program\'s version number and exit')
+
+    # configuration file
+    config_group = parser.add_argument_group('configuration arguments')
+    config_group.add_argument(
+        '-C', '--config', type=str, help='run with configuation file')
+    config_group.add_argument(
+        '-S',
+        '--section',
```
Comment (seanshpark) [inline compiler/one-cmds/one_build.py:43]: Do we really need this to get from the user?

File: compiler/one-cmds/one_build.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    """verify given arguments"""
+    # check if required arguments is given
+    if not _utils._is_valid_attr(args, 'config'):
+        parser.error('-C/--config argument is required')
+
+
+def _get_driver_name(driver_name):
+    return {
+        'one-import-bcq': 'one_import_bcq.py',
+        'one-import-tf': 'one_import_tf.py',
```
Comment (seanshpark) [inline compiler/one-cmds/one_build.py:69]: Why do you leave `one-import-tf` if you add `one_import_tf.py` ?

File: compiler/one-cmds/one_build.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        'one-import.tflite': 'one_import_tflite.py',
```
Comment (seanshpark) [inline compiler/one-cmds/one_build.py:70]: where did `one-import.tflite` came from? isn't it `one-import-`tflite ?

File: compiler/one-cmds/one_build.py (hunk 1 above)
Comment (mhs4670go) [reply compiler/one-cmds/one_build.py]: @seanshpark There are two options for users. If `--section` option is given, just run the section. If option is not given, all section in config file will be run sequentially.

Comment (seanshpark) [reply compiler/one-cmds/one_build.py]: > If option is not given, all section in config file will be run sequentially.

This means, config file is not just provides static information but a control information from `run sequentially.`.
I don't this is a good idea. It would be better not to include this implict behaviour.

Comment (seanshpark) [reply compiler/one-cmds/one_build.py]: In another point, users may want to get controllability of execution sequence somehow . Let's think about this...

**Comment by mhs4670go** [comment]
```bash
$ ./runtestall.sh
one-import-bcq_001.test SUCCESS
one-import-bcq_neg_001.test SUCCESS
one-import-bcq_neg_002.test SUCCESS
one-import-bcq_neg_003.test SUCCESS
one-import-bcq_neg_004.test SUCCESS
one-import-bcq_neg_005.test SUCCESS
one-import-bcq_neg_006.test SUCCESS
one-import-bcq_neg_007.test SUCCESS
one-import-bcq_neg_008.test SUCCESS
one-import-bcq_neg_009.test SUCCESS
one-import_001.test SUCCESS
one-import_neg_001.test SUCCESS
one-import_neg_002.test SUCCESS
one-import_neg_003.test SUCCESS
one-import_neg_004.test SUCCESS
one-import_neg_005.test SUCCESS
one-import_neg_006.test SUCCESS
one-import_neg_007.test SUCCESS
one-import_neg_008.test SUCCESS
one-import_neg_009.test SUCCESS
one-import_neg_010.test SUCCESS
one-optimize_001.test SUCCESS
one-optimize_neg_001.test SUCCESS
one-optimize_neg_002.test SUCCESS
one-optimize_neg_003.test SUCCESS
one-pack_001.test SUCCESS
one-pack_neg_001.test SUCCESS
one-pack_neg_002.test SUCCESS
one-pack_neg_003.test SUCCESS
one-quantize_001.test SUCCESS
one-quantize_neg_001.test SUCCESS
one-quantize_neg_002.test SUCCESS
one-quantize_neg_003.test SUCCESS
one-quantize_neg_004.test SUCCESS
one-quantize_neg_005.test SUCCESS
one-quantize_neg_006.test SUCCESS
one-quantize_neg_007.test SUCCESS
one-quantize_neg_008.test SUCCESS
one-quantize_neg_009.test SUCCESS
one-quantize_neg_010.test SUCCESS
one-quantize_neg_011.test SUCCESS
one-quantize_neg_012.test SUCCESS
one-quantize_neg_013.test SUCCESS
```

PTAL:)

File: compiler/bcq-tools/generate_bcq_output_arrays.py [hunk 4]
```diff
@@ -233,13 +233,13 @@ def print_bcqinfo_output_arrays_v1(flags):
                 f_arrays.write("," + prefix + "/bcqinfo_qbits_of_clusters")
                 f_arrays.write("," + fuseop)
                 if has_dequant_weight:
-                    f_arrays.write("," + prefix + "/bcqinfo_dequant_weight")
+                    f_arrays.write("," + prefix + "/bcqinfo_1dequant_weight")
```
Comment (seanshpark) [inline compiler/bcq-tools/generate_bcq_output_arrays.py:236]: there's a `1` in `bcqinfo_1dequant_weight`.

File: compiler/one-cmds/how-to-use-one-commands.txt [hunk 5]
```diff
@@ -20,6 +20,16 @@ Compilation flow for NPU
...
 
 
+Prerequisite
+------------
+
+Before you use one commands, please prepare a python environment with TensorFlow installed.
+You can use `one-prepare-venv`(ubuntu) script where it automatically installs python3 virtual
+environment with Tensorflow-cpu.
+
+If you have prepared Python as a virtual environment, don't forget to activate it.
```
Comment (seanshpark) [inline compiler/one-cmds/how-to-use-one-commands.txt:30]: Previouslly we didn't let users to activate but the shell script did it. Now we use python script and this makes activation not possible (or we don't know how-to).
What I could think of right now...
- add a script to activate and deactivate
- let one-import-tf as-is using bash
- ...

File: compiler/one-cmds/one-build.cfg [hunk 6]
```diff
@@ -0,0 +1,20 @@
+[driver-to-run]
+one-import-tf=True
+one-import-tflite=False
+one-import-bcq=False
+one-optimize=True
+one-quantize=False
+one-pack=False
```
Comment (seanshpark) [inline compiler/one-cmds/one-build.cfg:7]: Let's set to True

File: compiler/one-cmds/one-build.cfg [hunk 7]
```diff
@@ -0,0 +1,20 @@
+[driver-to-run]
```
Comment (seanshpark) [inline compiler/one-cmds/one-build.cfg:1]: How about naming this section as `one-build` ?

Comment (seanshpark) [inline compiler/one-cmds/one-build.cfg:1]: How about naming this filas as `one-build.cfg.template` ?

File: compiler/one-cmds/one-codegen [hunk 8]
```diff
@@ -14,53 +14,74 @@
...
+
+import one_utils as _utils
+
+
+def _get_parser():
+    parser = argparse.ArgumentParser(description='command line tool for code generation')
+
+    _utils._add_default_arg(parser)
+
+    parser.add_argument('-b', '--backend', type=str, help='backend name to use')
```
Comment (seanshpark) [inline compiler/one-cmds/one-codegen:31]: Please add the version option

File: compiler/one-cmds/one-build [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+def main():
+    # parse arguments
+    parser = _get_parser()
+    args = _parse_arg(parser)
+
+    # verify arguments
+    _verify_arg(parser, args)
+
+    # parse configuration file
+    config = _parse_cfg(args)
```
Comment (seanshpark) [inline compiler/one-cmds/one-build:96]: sequence of `_verify_arg` -> `_parse_cfg`: how does this work if both command line arg and cfg file exist?

File: compiler/one-cmds/one-codegen [hunk 10]
```diff
@@ -14,53 +14,74 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-DRIVER_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
-
-function Usage()
-{
-  echo "Usage: one-codegen [BACKEND] ..."
-  echo "Available BACKEND drivers:"
```
Comment (seanshpark) [inline compiler/one-cmds/one-codegen:22]: please add this part

**Comment by seanshpark** [comment]
Please add tests that uses `.cfg` files

File: compiler/bcq-tools/generate_bcq_output_arrays.py [hunk 11]
```diff
@@ -255,6 +255,196 @@ def print_bcq_output_arrays(flags):
         raise SystemExit(err_msg)
 
 
+def get_bcq_version(input_path):
```
Comment (seanshpark) [inline compiler/bcq-tools/generate_bcq_output_arrays.py:258]: Why is this added?

File: compiler/one-cmds/CMakeLists.txt [hunk 12]
```diff
@@ -8,6 +10,7 @@ set(ONE_COMMAND_FILES
     one-pack
     one-codegen
     one-prepare-venv
+    one_utils.py
```
Comment (seanshpark) [inline compiler/one-cmds/CMakeLists.txt:13]: Adding files in `ONE_COMMAND_FILES` will make it executable. Please add another list for `one-build.cfg` and `one_utils.py`.

Comment (seanshpark) [inline compiler/one-cmds/CMakeLists.txt:13]: following `one*` for `one_utils.py` seems confusing for command name similarity. can we use another name that doesn't start with `one` ?

File: compiler/one-cmds/tests/one-import-bcq_neg_003.test [hunk 13]
```diff
@@ -21,7 +21,7 @@ filename="${filename_ext%.*}"
 
 trap_err_onexit()
 {
-  if grep -q "input model not found" "${filename}.log"; then
```
Comment (seanshpark) [inline compiler/one-cmds/tests/one-import-bcq_neg_003.test:24]: This error message followed https://github.com/Samsung/ONE/issues/4155#issuecomment-690903095
@lemmaa , do we need to keep this or can we update this to follow new tools?

File: compiler/one-cmds/one_utils.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    parser.add_argument(
+        '-v',
+        '--version',
+        action='store_true',
+        help='show program\'s version number and exit')
+
+    # configuration file
+    parser.add_argument('-C', '--config', type=str, help='run with configuation file')
+    # section name that you want to run in configuration file
+    parser.add_argument('-S', '--section', type=str, help=argparse.SUPPRESS)
```
Comment (seanshpark) [inline compiler/one-cmds/one_utils.py:35]: do we need this? as we discussed off-line, we could remove this for the first version.

File: compiler/bcq-tools/generate_bcq_output_arrays.py (hunk 11 above)
Comment (mhs4670go) [reply compiler/bcq-tools/generate_bcq_output_arrays.py:258]: @seanshpark print_* method will be deprecated. Because it returns its output as a file, which is redundant if we use python script. So, get_* methods return its output as python things.

File: compiler/one-cmds/one-build (hunk 9 above)
Comment (mhs4670go) [reply compiler/one-cmds/one-build:96]: @seanshpark As we've discussed earlier, arg overrides cfg.

File: compiler/one-cmds/one_utils.py (hunk 14 above)
Comment (mhs4670go) [reply compiler/one-cmds/one_utils.py:35]: @seanshpark if `argparse.SUPPRESS` is given to `help` kwarg, user can't see its option. It is internal or private option for implementation convenience and scalability.

File: compiler/one-cmds/one-build (hunk 9 above)
Comment (seanshpark) [reply compiler/one-cmds/one-build:96]: Why I asked is funcion call sequence is `_verify_arg` then `_parse_cfg`.
One may think (1) set the `args` with command line (2) override `args` with cfg file.
So, can we swap these?

File: compiler/bcq-tools/generate_bcq_output_arrays.py [hunk 15]
```diff
@@ -255,6 +255,196 @@ def print_bcq_output_arrays(flags):
...
+            ret_output_arrays.append(prefix + '/bcqinfo_size_of_clusters')
+            ret_output_arrays.append(prefix + '/bcqinfo_qbits_of_clusters')
+            ret_output_arrays.append(fuseop)
+            if has_dequant_weight:
+                ret_output_arrays.append(prefix + '/bcqinfo_dequant_weight')
+
+    return ret_output_arrays
+
+
+def get_bcq_output_arrays(input_path, output_arrays):
```
Comment (seanshpark) [inline compiler/bcq-tools/generate_bcq_output_arrays.py:432]: > Why is this added?

Maybe I should have asked here...  this file is in `bcq-tools` not in `one-cmds`

File: compiler/one-cmds/one_utils.py (hunk 14 above)
Comment (seanshpark) [reply compiler/one-cmds/one_utils.py:35]: Can user add this in argument and override it? Users can still access the source code.

File: compiler/one-cmds/one-build (hunk 9 above)
Comment (mhs4670go) [reply compiler/one-cmds/one-build:96]: Unfortunately, I can't. Actually, I have to verify `config` - configuration file path - argument  before parsing configuration file. It seems natural...

Comment (seanshpark) [reply compiler/one-cmds/one-build:96]: Sorry, not `_verify_arg`...
it is about `_parse_arg` -> `_parse_cfg` sequence.

Comment (mhs4670go) [reply compiler/one-cmds/one-build:96]: @seanshpark Right. It might seem unnatural as you said. But as with above, conf file path first has to be given before parsing conf file. So I should parse arg first.

File: compiler/one-cmds/one_utils.py (hunk 14 above)
Comment (mhs4670go) [reply compiler/one-cmds/one_utils.py:35]: Yes. Umm.. I think there is no demerit even if user finds out this option from source code. If you worries about misbehave like randomly user gives `-S`, we can print warning at this time.

File: compiler/bcq-tools/generate_bcq_output_arrays.py (hunk 15 above)
Comment (mhs4670go) [reply compiler/bcq-tools/generate_bcq_output_arrays.py:432]: Oh, right. I will split this file when I post a PR.

File: compiler/one-cmds/tests/one-import-bcq_neg_003.test (hunk 13 above)
Comment (seanshpark) [reply compiler/one-cmds/tests/one-import-bcq_neg_003.test:24]: Had a off-line talk with @lemma and it is OK to change these :)

File: compiler/one-cmds/one_utils.py (hunk 14 above)
Comment (seanshpark) [reply compiler/one-cmds/one_utils.py:35]: OK, got it. I think no need to add warning message. Using -`S` is possible but it's up to the user to use unofficial arguments and get misbehaviour.

File: compiler/one-cmds/one-build.cfg (hunk 7 above)
Comment (mhs4670go) [reply compiler/one-cmds/one-build.cfg:1]: @seanshpark How about `one-build.template.cfg`? Editor programs like vim or sth recognize its format as a configuration  file from last extension. 
`one-build.cfg.template`
image
`one-build.template.cfg`
image

File: compiler/one-cmds/how-to-use-one-commands.txt (hunk 5 above)
Comment (mhs4670go) [reply compiler/one-cmds/how-to-use-one-commands.txt:30]: @seanshpark I used this magic:)
```bash
#!/usr/bin/env bash
''''export SCRIPT_PATH="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"                  # '''
''''export PY_PATH=${SCRIPT_PATH}/venv/bin/python                                       # '''
''''test -f ${PY_PATH} && exec ${PY_PATH} "$0" "$@"                                     # '''
''''echo "Error: Virtual environment not found. Please run 'one-prepare-venv' command." # '''
```
First, it will be run as a bash script. And then, check if python path exists. If it does, execute it. Triple quote will be comment then.

File: compiler/one-cmds/CMakeLists.txt [hunk 16]
```diff
@@ -36,6 +34,34 @@ foreach(ONE_COMMAND IN ITEMS ${ONE_COMMAND_FILES})
           
 endforeach(ONE_COMMAND)
 
+set(ONE_UTILITIY_FIELS
```
Comment (seanshpark) [inline compiler/one-cmds/CMakeLists.txt:37]: typo `ONE_UTILITIY_FILES` ?

Comment (mhs4670go) [reply compiler/one-cmds/CMakeLists.txt:37]: Oh, Hawk's eye!

File: compiler/one-cmds/CMakeLists.txt [hunk 17]
```diff
@@ -34,17 +34,17 @@ foreach(ONE_COMMAND IN ITEMS ${ONE_COMMAND_FILES})
           
 endforeach(ONE_COMMAND)
 
-set(ONE_UTILITIY_FIELS
+set(ONE_UTILITY_FIELS
```
Comment (seanshpark) [inline compiler/one-cmds/CMakeLists.txt:37]: You've fixed to remove `I` in `UTILITY`- you have good eyes too
What I saw `ONE_UTILITY_FILES` in `FIELS` -> `FILES`

File: compiler/one-cmds/how-to-use-one-commands.txt (hunk 5 above)
Comment (seanshpark) [reply compiler/one-cmds/how-to-use-one-commands.txt:30]: This is very interesting!

Comment (mhs4670go) [reply compiler/one-cmds/how-to-use-one-commands.txt:30]: @seanshpark Actually, there is pros and cons.
`one-import-tf.py` + `one-import` = increase the number of files
`one-import-tf`(bash + python as above) = need manual format check and no ide formatting
